// ngx-img/esm2015/module/component/ngx-img.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, EventEmitter, Input, OnChanges, OnDestroy, OnInit, Output, SimpleChanges, ViewChild } from '@angular/core';
import { NgxImgService } from '../service/ngx-img.service';
let NgxImgComponent = class NgxImgComponent {
    constructor(_service, _ref) {
        this._service = _service;
        this._ref = _ref;
        this.alt = '';
        this.fileName = '';
        this.imgSrc = '';
        this.remove = true;
        this.hasPreview = false;
        this.hasError = false;
        this.isLoading = false;
        this._config = {
            fileSize: 2048,
            minWidth: 0,
            maxWidth: 0,
            minHeight: 0,
            maxHeight: 0,
            fileType: ['.gif', '.jpeg', '.png', '.jpg'],
            quality: 0.8
        };
        this._text = {
            default: 'Drag and drop',
            _default: 'Drag and drop or click',
            button: 'Choose File',
            try_again: 'Try Again',
            replace: 'Drag and drop or click to replace',
            reset: 'Remove',
            error: 'Oops, something wrong happened.'
        };
        this._errorTexts = {
            fileSize: 'The file size is too big ({{ value }} max).',
            minWidth: 'The image width is too small ({{ value }}}px min).',
            maxWidth: 'The image width is too big ({{ value }}}px max).',
            minHeight: 'The image height is too small ({{ value }}}px min).',
            maxHeight: 'The image height is too big ({{ value }}}px max).',
            imageFormat: 'The image format is not allowed ({{ value }} only).',
            fileType: 'The file type is not allowed.'
        };
        this.errors = [];
        this.mode = 'upload';
        this.onChange = new EventEmitter();
        this.onSelect = new EventEmitter();
        this.onReset = new EventEmitter();
        this.reset();
    }
    ngOnInit() {
        this._text = Object.assign(this._text, this.text);
        this._errorTexts = Object.assign(this._errorTexts, this.errorTexts);
        this._config = Object.assign(this._config, this.config);
        this.markForCheck();
    }
    ngOnChanges(changes) {
        if (changes.imgSrc && changes.imgSrc.isFirstChange()) {
            if (this.imgSrc) {
                this.hasPreview = true;
                if (this.config && this.config.crop) {
                    this.mode = 'crop';
                }
                this.markForCheck();
            }
        }
    }
    fileChangeListener(e) {
        this.hasError = false;
        this.errors = [];
        if (!e.target.files.length) {
            this.reset();
            return false;
        }
        this.file = e.target.files[0];
        if (!this.validate()) {
            this.hasError = true;
            this.reset();
            return false;
        }
        this.fileName = this.file.name;
        this.onChange.emit(this.file);
        if (this.file.type.split('/')[0] === 'image') {
            this.isLoading = true;
            const reader = new FileReader();
            reader.onloadend = (ev) => {
                this.imgSrc = ev.target.result;
                this.fileName = this.file.name;
                this.hasPreview = true;
                this.isLoading = false;
                if (this._config.crop) {
                    this.mode = 'crop';
                    this.markForCheck();
                }
                else {
                    this._service.compress(this.imgSrc, this._config).then((res) => {
                        this.onSelectEvent(res);
                        this.markForCheck();
                    })
                        .catch(() => {
                        this.onSelectEvent(this.imgSrc);
                        this.markForCheck();
                    });
                }
            };
            reader.readAsDataURL(this.file);
            this.markForCheck();
        }
        this.markForCheck();
    }
    reset() {
        this.mode = 'upload';
        this.file = null;
        this.imgSrc = '';
        this.fileName = '';
        this.hasPreview = false;
        if (this.fileInput) {
            this.fileInput.nativeElement.value = '';
        }
        this.onReset.emit();
        this.markForCheck();
    }
    validate() {
        this.errors = [];
        if (this._config.fileType.length !== 0 &&
            this._config.fileType.indexOf(this.file.type) === -1 && this._config.fileType.indexOf(`.${this.file.name.split('.').pop()}`) === -1) {
            this.errors = [...this.errors, this._errorTexts.fileType];
        }
        if (this._config.fileSize !== 0 && (this.file.size / 1024) > this._config.fileSize) {
            this.errors = [...this.errors, this._errorTexts.fileSize.replace('{{ value }}', this.sizeToByte(this._config.fileSize))];
        }
        if (this._config.minWidth !== 0 && this._config.minWidth >= this.file.width) {
            this.errors = [...this.errors, this._errorTexts.minWidth.replace('{{ value }}', this._config.minWidth.toString())];
        }
        if (this._config.maxWidth !== 0 && this._config.maxWidth <= this.file.width) {
            this.errors = [...this.errors, this._errorTexts.maxWidth.replace('{{ value }}', this._config.maxWidth.toString())];
        }
        if (this._config.minHeight !== 0 && this._config.minHeight >= this.file.height) {
            this.errors = [...this.errors, this._errorTexts.minHeight.replace('{{ value }}', this._config.minHeight.toString())];
        }
        if (this._config.maxHeight !== 0 && this._config.maxHeight <= this.file.height) {
            this.errors = [...this.errors, this._errorTexts.maxHeight.replace('{{ value }}', this._config.maxHeight.toString())];
        }
        return !this.errors.length;
    }
    sizeToByte(size) {
        const units = ['KB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB'];
        let i = 0;
        while (size >= 1024) {
            size /= 1024;
            ++i;
        }
        return size.toFixed(1) + ' ' + units[i];
    }
    ;
    onSelectEvent(data) {
        this.onSelect.emit(data);
        this.markForCheck();
    }
    markForCheck() {
        setTimeout(() => {
            if (!this._ref['destroyed']) {
                this._ref.markForCheck();
                this._ref.detectChanges();
            }
        });
        setTimeout(() => {
            if (!this._ref['destroyed']) {
                this._ref.markForCheck();
                this._ref.detectChanges();
            }
        }, 300);
    }
    ngOnDestroy() {
        this.reset();
        this._ref.detach();
    }
};
NgxImgComponent.ctorParameters = () => [
    { type: NgxImgService },
    { type: ChangeDetectorRef }
];
__decorate([
    Input()
], NgxImgComponent.prototype, "alt", void 0);
__decorate([
    Input()
], NgxImgComponent.prototype, "fileName", void 0);
__decorate([
    Input()
], NgxImgComponent.prototype, "imgSrc", void 0);
__decorate([
    Input()
], NgxImgComponent.prototype, "remove", void 0);
__decorate([
    ViewChild('fileInput')
], NgxImgComponent.prototype, "fileInput", void 0);
__decorate([
    Input()
], NgxImgComponent.prototype, "config", void 0);
__decorate([
    Input()
], NgxImgComponent.prototype, "errorTexts", void 0);
__decorate([
    Input()
], NgxImgComponent.prototype, "text", void 0);
__decorate([
    Output()
], NgxImgComponent.prototype, "onChange", void 0);
__decorate([
    Output()
], NgxImgComponent.prototype, "onSelect", void 0);
__decorate([
    Output()
], NgxImgComponent.prototype, "onReset", void 0);
NgxImgComponent = __decorate([
    Component({
        selector: 'ngx-img',
        template: "<ng-container [ngSwitch]=\"mode\">\n  <ng-template [ngSwitchCase]=\"'upload'\">\n    <div class=\"ngx-img-wrapper\" [class.has-preview]=\"hasPreview\" [class.has-error]=\"hasError\"\n         [style.height]=\"_config.height\">\n      <div class=\"ngx-img-message\" *ngIf=\"!hasPreview\">\n        <svg xmlns=\"http://www.w3.org/2000/svg\" width=\"60\" height=\"60\" viewBox=\"0 0 640 640\" class=\"ngx-img-upload\">\n          <path class=\"st-1\"\n                d=\"M411.312 372.688l-64-64c-6.247-6.248-16.379-6.248-22.627 0l-64 64c-6.248 6.247-6.248 16.379 0 22.627s16.379 6.248 22.628 0l36.687-36.687v137.372c0 8.837 7.163 16 16 16s16-7.163 16-16v-137.372l36.688 36.687c3.123 3.123 7.217 4.685 11.312 4.685s8.189-1.562 11.312-4.688c6.25-6.247 6.25-16.377 0-22.624z\"></path>\n          <path class=\"st-2\"\n                d=\"M512.2 512h-112.2c-8.837 0-16-7.163-16-16s7.163-16 16-16h112.2c52.824 0 95.8-42.976 95.8-95.8s-42.976-95.8-95.8-95.8c-11.33 0-22.371 1.938-32.817 5.764-6.979 2.555-14.798-0.031-18.876-6.243s-3.343-14.415 1.776-19.803c11.425-12.022 17.717-27.692 17.717-44.118 0-35.29-28.71-64-64-64-20.536 0-39.32 9.492-51.535 26.042-7.374 9.991-11.668 21.692-12.417 33.836-0.419 6.781-5.073 12.558-11.609 14.409-6.539 1.854-13.53-0.629-17.441-6.182-3.52-4.998-7.331-9.814-11.329-14.314-30.367-34.185-73.984-53.791-119.669-53.791-88.224 0-160 71.775-160 160s71.776 160 160 160h80c8.836 0 16 7.163 16 16s-7.164 16-16 16h-80c-105.869 0-192-86.131-192-192s86.131-192 192-192c51.271 0 100.368 20.573 136.315 56.791 2.802-6.217 6.28-12.168 10.402-17.752 18.044-24.445 46.933-39.039 77.283-39.039 52.935 0 96 43.065 96 96 0 11.223-1.949 22.202-5.685 32.532 1.954-0.089 3.917-0.132 5.885-0.132 70.469 0 127.8 57.331 127.8 127.8s-57.331 127.8-127.8 127.8z\"></path>\n        </svg>\n        <ng-template [ngIf]=\"!hasError\" [ngIfElse]=\"withErrorBlock\">\n          <p>{{ _text.default }}</p>\n          <button class=\"ngx-img-button\">{{ _text.button }}</button>\n        </ng-template>\n        <ng-template #withErrorBlock>\n          <p>{{ _text._default }}</p>\n          <div class=\"ngx-img-error\" *ngIf=\"hasError\">\n            <ul>\n              <li *ngFor=\"let error of errors\">{{ error }}</li>\n            </ul>\n          </div>\n          <button style=\"margin-top: 0\" class=\"ngx-img-button\">{{ _text.try_again }}</button>\n        </ng-template>\n      </div>\n      <div class=\"ngx-img-loader\" *ngIf=\"isLoading\"></div>\n      <input type=\"file\" [accept]=\"_config.fileType.join(', ')\" #fileInput (change)=\"fileChangeListener($event)\">\n      <button *ngIf=\"remove && hasPreview\" type=\"button\" class=\"ngx-img-clear\" (click)=\"reset()\">{{ _text.reset }}\n      </button>\n      <div class=\"ngx-img-preview\" *ngIf=\"hasPreview\">\n        <span class=\"ngx-img-render\"><img [src]=\"imgSrc\" [alt]=\"alt\"></span>\n        <div class=\"ngx-img-infos\">\n          <div class=\"ngx-img-infos-inner\">\n            <p class=\"ngx-img-filename\">{{fileName}}</p>\n            <p class=\"ngx-img-infos-message\">{{ _text.replace }}</p>\n          </div>\n        </div>\n      </div>\n    </div>\n  </ng-template>\n  <ng-template [ngSwitchCase]=\"'crop'\">\n    <ngx-img-crop [config]=\"config\" [imgSrc]=\"imgSrc\" (onCrop)=\"onSelectEvent($event)\"\n                  (onReset)=\"reset()\"></ngx-img-crop>\n  </ng-template>\n</ng-container>\n",
        changeDetection: ChangeDetectionStrategy.OnPush,
        styles: [".ngx-img-wrapper{display:block;position:relative;cursor:pointer;overflow:hidden;width:100%;max-width:100%;height:400px;padding:5px 10px;font-family:Helvetica,Arial,sans-serif;font-size:14px;font-weight:300;line-height:22px;color:#333;background-color:#fff;background-image:none;text-align:center;border:2px solid #e5e5e5;transition:border-color .15s linear}.ngx-img-wrapper:hover{background-size:30px 30px;background-image:linear-gradient(-45deg,#f6f6f6 25%,transparent 25%,transparent 50%,#f6f6f6 50%,#f6f6f6 75%,transparent 75%,transparent);-webkit-animation:2s linear infinite stripes;animation:2s linear infinite stripes}.ngx-img-wrapper.has-preview .ngx-img-clear{display:block}.ngx-img-wrapper.has-error{border-color:#f44336}.ngx-img-wrapper.has-error .ngx-img-message .ngx-img-error{display:block}.ngx-img-wrapper.has-error:hover .ngx-img-errors-container{visibility:visible;opacity:1;transition-delay:0s}.ngx-img-wrapper.disabled input{cursor:not-allowed}.ngx-img-wrapper.disabled:hover{background-image:none;-webkit-animation:none;animation:none}.ngx-img-wrapper.disabled .ngx-img-message{opacity:.5;text-decoration:line-through}.ngx-img-wrapper.disabled .ngx-img-infos-message{display:none}.ngx-img-wrapper input{position:absolute;top:0;right:0;bottom:0;left:0;height:100%;width:100%;opacity:0;cursor:pointer;z-index:5}.ngx-img-wrapper .ngx-img-button{font-size:.7rem;padding:.6rem 1rem;background:#17a2c4;box-shadow:none;border-radius:10em;border:0;color:#fff;margin-top:1rem;text-transform:uppercase;display:inline-block;font-weight:400;line-height:1.25;text-align:center;white-space:nowrap;vertical-align:middle;cursor:pointer;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none}.ngx-img-wrapper .ngx-img-message{position:relative;top:50%;transform:translateY(-50%)}.ngx-img-wrapper .ngx-img-message svg.ngx-img-upload .st-1,.ngx-img-wrapper .ngx-img-message svg.ngx-img-upload .st-2{fill:#17a2c4}.ngx-img-wrapper .ngx-img-message div,.ngx-img-wrapper .ngx-img-message p{margin:5px 0 0}.ngx-img-wrapper .ngx-img-message div.ngx-img-error,.ngx-img-wrapper .ngx-img-message p.ngx-img-error{color:#f44336;font-weight:700;display:none}.ngx-img-wrapper .ngx-img-message div li,.ngx-img-wrapper .ngx-img-message div ul,.ngx-img-wrapper .ngx-img-message p li,.ngx-img-wrapper .ngx-img-message p ul{list-style:none;padding:0}.ngx-img-wrapper .ngx-img-clear{display:block;position:absolute;opacity:0;z-index:7;top:10px;right:10px;background:0 0;border:2px solid #fff;text-transform:uppercase;font-family:Helvetica,Arial;font-size:11px;padding:4px 8px;font-weight:700;color:#fff;transition:.15s linear}.ngx-img-wrapper .ngx-img-clear:hover{background:rgba(255,255,255,.2)}.ngx-img-wrapper .ngx-img-preview{display:block;position:absolute;z-index:1;background-color:#fff;padding:5px;width:100%;height:100%;top:0;right:0;bottom:0;left:0;overflow:hidden;text-align:center}.ngx-img-wrapper .ngx-img-preview .ngx-img-render img{top:50%;transform:translate(0,-50%);position:relative;max-width:100%;max-height:100%;background-color:#fff;transition:border-color .15s linear}.ngx-img-wrapper .ngx-img-preview .ngx-img-render i{font-size:70px;top:50%;left:50%;transform:translate(-50%,-50%);position:absolute;color:#777}.ngx-img-wrapper .ngx-img-preview .ngx-img-render .ngx-img-extension{position:absolute;top:50%;left:50%;transform:translate(-50%,-50%);margin-top:10px;text-transform:uppercase;font-weight:900;letter-spacing:-.03em;font-size:13px;width:42px;white-space:nowrap;overflow:hidden;text-overflow:ellipsis}.ngx-img-wrapper .ngx-img-preview .ngx-img-infos{position:absolute;left:0;top:0;right:0;bottom:0;z-index:3;background:rgba(0,0,0,.7);opacity:0;transition:opacity .15s linear}.ngx-img-wrapper .ngx-img-preview .ngx-img-infos .ngx-img-infos-inner{position:absolute;top:50%;transform:translate(0,-40%);-webkit-backface-visibility:hidden;backface-visibility:hidden;width:100%;padding:0 20px;transition:.2s}.ngx-img-wrapper .ngx-img-preview .ngx-img-infos .ngx-img-infos-inner p{padding:0;margin:0;position:relative;width:100%;white-space:nowrap;overflow:hidden;text-overflow:ellipsis;color:#fff;text-align:center;line-height:25px;font-weight:700}.ngx-img-wrapper .ngx-img-preview .ngx-img-infos .ngx-img-infos-inner p.ngx-img-infos-message{margin-top:15px;padding-top:15px;font-size:12px;position:relative;opacity:.5}.ngx-img-wrapper .ngx-img-preview .ngx-img-infos .ngx-img-infos-inner p.ngx-img-infos-message::before{content:\"\";position:absolute;top:0;left:50%;transform:translate(-50%,0);background:#fff;width:30px;height:2px}.ngx-img-wrapper:hover .ngx-img-clear,.ngx-img-wrapper:hover .ngx-img-preview .ngx-img-infos{opacity:1}.ngx-img-wrapper:hover .ngx-img-preview .ngx-img-infos .ngx-img-infos-inner{margin-top:-5px}.ngx-img-wrapper.touch-fallback{height:auto!important}.ngx-img-wrapper.touch-fallback:hover{background-image:none;-webkit-animation:none;animation:none}.ngx-img-wrapper.touch-fallback .ngx-img-preview{position:relative;padding:0;display:block}.ngx-img-wrapper.touch-fallback .ngx-img-preview .ngx-img-render{display:block;position:relative}.ngx-img-wrapper.touch-fallback .ngx-img-preview .ngx-img-render img{position:relative;transform:translate(0,0)}.ngx-img-wrapper.touch-fallback .ngx-img-preview .ngx-img-infos{position:relative;opacity:1;background:0 0}.ngx-img-wrapper.touch-fallback .ngx-img-preview .ngx-img-infos .ngx-img-infos-inner{position:relative;top:0;transform:translate(0,0);padding:5px 90px 5px 0}.ngx-img-wrapper.touch-fallback .ngx-img-preview .ngx-img-infos .ngx-img-infos-inner p{padding:0;margin:0;position:relative;width:100%;white-space:nowrap;overflow:hidden;text-overflow:ellipsis;color:#777;text-align:left;line-height:25px}.ngx-img-wrapper.touch-fallback .ngx-img-preview .ngx-img-infos .ngx-img-infos-inner p.ngx-img-filename{font-weight:700}.ngx-img-wrapper.touch-fallback .ngx-img-preview .ngx-img-infos .ngx-img-infos-inner p.ngx-img-infos-message{margin-top:0;padding-top:0;font-size:11px;position:relative;opacity:1}.ngx-img-wrapper.touch-fallback .ngx-img-message{transform:translate(0,0);padding:40px 0}.ngx-img-wrapper.touch-fallback .ngx-img-clear{top:auto;bottom:23px;opacity:1;border-color:rgba(119,119,119,.7);color:#777}.ngx-img-wrapper.touch-fallback.has-preview .ngx-img-message{display:none}.ngx-img-wrapper.touch-fallback:hover .ngx-img-preview .ngx-img-infos .ngx-img-infos-inner{margin-top:0}.ngx-img-wrapper .ngx-img-loader{position:absolute;top:15px;right:15px;display:block;z-index:9}.ngx-img-wrapper .ngx-img-loader::after{display:block;position:relative;width:20px;height:20px;-webkit-animation:.6s linear infinite rotate;animation:.6s linear infinite rotate;border-radius:100%;border-top:1px solid #ccc;border-bottom:1px solid #777;border-left:1px solid #ccc;border-right:1px solid #777;content:\"\"}.ngx-img-wrapper .ngx-img-errors-container{position:absolute;left:0;top:0;right:0;bottom:0;z-index:3;background:rgba(244,67,54,.8);text-align:left;visibility:hidden;opacity:0;transition:visibility linear .15s,opacity .15s linear}.ngx-img-wrapper .ngx-img-errors-container ul{padding:10px 20px;margin:0;position:absolute;left:0;top:50%;transform:translateY(-50%)}.ngx-img-wrapper .ngx-img-errors-container ul li{margin-left:20px;color:#fff;font-weight:700}.ngx-img-wrapper .ngx-img-errors-container.visible{visibility:visible;opacity:1;transition-delay:0s}.ngx-img-wrapper~.ngx-img-errors-container ul{padding:0;margin:15px 0}.ngx-img-wrapper~.ngx-img-errors-container ul li{margin-left:20px;color:#f44336;font-weight:700}@-webkit-keyframes stripes{from{background-position:0 0}to{background-position:60px 30px}}@keyframes stripes{from{background-position:0 0}to{background-position:60px 30px}}@-webkit-keyframes rotate{0%{transform:rotateZ(-360deg)}100%{transform:rotateZ(0)}}@keyframes rotate{0%{transform:rotateZ(-360deg)}100%{transform:rotateZ(0)}}"]
    })
], NgxImgComponent);
export { NgxImgComponent };
