// ngx-img/esm5/module/ngx-img.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { NgxImgComponent } from './component/ngx-img.component';
import { NgxImgCropComponent } from './component/ngx-img-crop.component';
import { NgxImgService } from './service/ngx-img.service';
// Export module's public API
export { NgxImgComponent } from './component/ngx-img.component';
export { NgxImgCropComponent } from './component/ngx-img-crop.component';
export { NgxImgService } from './service/ngx-img.service';
var NgxImgModule = /** @class */ (function () {
    function NgxImgModule() {
    }
    NgxImgModule_1 = NgxImgModule;
    NgxImgModule.forRoot = function () {
        return {
            ngModule: NgxImgModule_1,
            providers: [NgxImgService]
        };
    };
    var NgxImgModule_1;
    NgxImgModule = NgxImgModule_1 = __decorate([
        NgModule({
            imports: [
                CommonModule,
                FormsModule
            ],
            exports: [NgxImgComponent, NgxImgCropComponent],
            declarations: [NgxImgComponent, NgxImgCropComponent],
            providers: [NgxImgService]
        })
    ], NgxImgModule);
    return NgxImgModule;
}());
export { NgxImgModule };
